database/plugin/metadata/sqlite/transaction.go

```diff
@@ -288,6 +288,130 @@ func (d *MetadataStoreSqlite) SetTransaction(
 			return result.Error
 		}
 	}
+	// Extract and save witness set data
+	// Delete existing witness records to ensure idempotency on retry
+	if tmpTx.ID != 0 {
+		if result := txn.Where("transaction_id = ?", tmpTx.ID).Delete(&models.KeyWitness{}); result.Error != nil {
+			return fmt.Errorf("delete existing key witnesses: %w", result.Error)
+		}
+		if result := txn.Where("transaction_id = ?", tmpTx.ID).Delete(&models.Script{}); result.Error != nil {
+			return fmt.Errorf("delete existing scripts: %w", result.Error)
+		}
+		if result := txn.Where("transaction_id = ?", tmpTx.ID).Delete(&models.Redeemer{}); result.Error != nil {
+			return fmt.Errorf("delete existing redeemers: %w", result.Error)
+		}
+		if result := txn.Where("transaction_id = ?", tmpTx.ID).Delete(&models.PlutusData{}); result.Error != nil {
+			return fmt.Errorf("delete existing plutus data: %w", result.Error)
+		}
+	}
```

Comment:
Instead of deleting existing records, we should probably use `ON CONFLICT` clauses to update the existing record(s)